### PR TITLE
Bundle

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: src/build/bundle.ts [hunk 1]
```diff
@@ -9,55 +9,328 @@
...
 
 import {Logger} from './logger';
 import {compose} from './streams';
 import {StreamResolver} from './stream-resolver';
 import {VulcanizeTransform} from './vulcanize';
+import {Analyzer, Loader} from 'hydrolysis';
 
 // non-ES module
 const minimatchAll = require('minimatch-all');
+const through = require('through2').obj;
```
Comment (azakus) [inline src/build/bundle.ts]: Let's just use a Transform stream.

File: src/build/bundle.ts [hunk 2]
```diff
@@ -9,55 +9,328 @@
...
+    this.entrypointFiles = new Map();
+
     this.bundle = compose([
       this.streamResolver,
-      gulpif((file) => this.isEntryPoint(file.path), this.vulcanize)
+      through((file, enc, cb) => {
+        if (this.isEntryPoint(file.path)) {
+          // don't pass on any entrypoints until the stream has ended
+          this.entrypointFiles.set(file.path, file);
+          cb(null, null);
```
Comment (azakus) [inline src/build/bundle.ts]: Let's check that `cb(null, null)` doesn't end the stream

File: src/build/bundle.ts [hunk 3]
```diff
@@ -9,55 +9,328 @@
...
+      // We want to collect dependencies that appear in > 1 entrypoint, but
+      // we need to collect them in document order, so rather than iterate
+      // directly through each dependency in depsToEntrypoints, we iterate
+      // through entrypoints in entrypointToDeps, which has dependencies in
+      // order for each entrypoint. Then we iterate through dependencies for
+      // each entrypoint and look up how many entrypoints depend on it.
+      // This assumes an ordering between entrypoints, since they could have
+      // conflicting orders between their top level imports. The shell should
+      // always come first.
+      for (let entrypoint of entrypointToDeps.keys()) {
```
Comment (azakus) [inline src/build/bundle.ts]: Maybe `let [entrypoint, dependencies] of entrypointToDeps.entries()`?

File: src/build/stream-resolver.ts [hunk 4]
```diff
@@ -141,11 +141,22 @@ export class StreamResolver extends Transform /* implements Resolver */ {
 
   end() {
     // When the stream is done, reject all non-resolved requests for files
-    for (let deferred of this._deferreds.values()) {
-      deferred.reject(null);
+    for (let url of this._deferreds.keys()) {
```
Comment (azakus) [inline src/build/stream-resolver.ts]: `let [url, deferred] of this._deferreds.entries()`

**Comment by azakus** [comment]
LGTM with nits

File: src/build/bundle.ts (hunk 3 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: I've been hesitant to use destructuring, given it's spotty availability

Comment (azakus) [inline src/build/bundle.ts]: SGTM

File: src/build/bundle.ts (hunk 2 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: I'm confident it doesn't, given that the entrypoints are eventually written out. This is covered in the tests.

File: src/build/bundle.ts (hunk 1 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: I'm going to merge as-is, this PR is needed asap

**Comment by garlicnation** [comment]
I'd like to see a test of document ordering for `_getBundles`. Otherwise LGTM.

**Pull request merged**